### PR TITLE
Feature/faktor export

### PR DESCRIPTION
**Pull request by tuj**
* Adds basic setup with api-platform, mongodb and an entry document.
* Adds command to extract stats from elasticsearch that should be delivered to Faktor.

Depends on https://github.com/danskernesdigitalebibliotek/ddb-cover-service/pull/53 being merged.

--------------------------------

Code Coverage Report:      
  2020-01-22 14:16:20      
                           
 Summary:                  
  Classes: 54.55% (6/11)   
  Methods: 88.00% (44/50)  
  Lines:   95.75% (248/259)

\App\Command::App\Command\CleanupEntriesCommand
  Methods: 100.00% ( 3/ 3)   Lines: 100.00% ( 12/ 12)
\App\Command::App\Command\ExtractStatisticsCommand
  Methods: 100.00% ( 3/ 3)   Lines: 100.00% (  9/  9)
\App\Command::App\Command\FakeCommand
  Methods:  66.67% ( 2/ 3)   Lines:  86.67% ( 13/ 15)
\App\Document::App\Document\Entry
  Methods: 100.00% (21/21)   Lines: 100.00% ( 31/ 31)
\App\Document::App\Document\ExtractionResult
  Methods:  60.00% ( 3/ 5)   Lines:  71.43% (  5/  7)
\App\EventSubscriber::App\EventSubscriber\ResponseSubscriber
  Methods: 100.00% ( 3/ 3)   Lines: 100.00% ( 21/ 21)
\App\Repository::App\Repository\EntryRepository
  Methods: 100.00% ( 2/ 2)   Lines: 100.00% (  5/  5)
\App\Repository::App\Repository\ExtractionResultRepository
  Methods: 100.00% ( 2/ 2)   Lines: 100.00% (  9/  9)
\App\Service::App\Service\ElasticsearchService
  Methods:  50.00% ( 1/ 2)   Lines:  95.83% ( 23/ 24)
\App\Service::App\Service\StatisticsExtractionService
  Methods:  75.00% ( 3/ 4)   Lines:  95.00% ( 95/100)
\App\Test::App\Test\DataFakerService
  Methods:  50.00% ( 1/ 2)   Lines:  96.15% ( 25/ 26)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,132 @@
+{
+    "name": "danskernesdigitalebibliotek/ddb-cover-service-faktor-export",
+    "description": "Data extraction API for Faktor.",
+    "type": "project",
+    "license": "proprietary",
```
Comment (cableman) [inline composer.json]: Maybe added "GNU AFFERO V3"

File: config/packages/api_platform.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        <ul>
+        <li>Oauth: not implemented.</li>
+        </ul>
+        <p><a href="https://www.gnu.org/licenses/agpl-3.0.en.html">GNU Affero General Public License</a></p>
+    show_webby: false
+    enable_re_doc: false
+    formats:
+        json:
+            mime_types: ['application/json']
+        jsonld:
```
Comment (cableman) [inline config/packages/api_platform.yaml]: Maybe remove jsonld as formart, mostly because the other services don't have it.

File: config/packages/cache.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+framework:
+    cache:
```
Comment (cableman) [inline config/packages/cache.yaml:2]: Please see https://github.com/danskernesdigitalebibliotek/ddb-cover-service/blob/develop/config/packages/framework.yaml#L17 to cache symfony into redis... DB 12 is free (https://github.com/danskernesdigitalebibliotek/ddb-cover-service/blob/develop/.env#L37).

File: config/routes/api_platform.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+api_platform:
+    resource: .
+    type: api_platform
+    prefix: /
```
Comment (cableman) [inline config/routes/api_platform.yaml]: The other all have the prefix /api and an route redirect from `/` to `/api` ?

Comment (cableman) [reply config/routes/api_platform.yaml]: https://github.com/danskernesdigitalebibliotek/ddb-cover-service/blob/develop/config/routes.yaml

File: config/packages/security.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    providers:
+        in_memory: { memory: null }
+    firewalls:
+        dev:
+            pattern: ^/(_(profiler|wdt)|css|images|js)/
+            security: false
+        main:
+            anonymous: lazy
+
+            # activate different ways to authenticate
```
Comment (cableman) [inline config/packages/security.yaml:12]: In CoverService we have disabled security for the api parts to enhance performance: https://github.com/danskernesdigitalebibliotek/ddb-cover-service/blob/develop/config/packages/security.yaml#L9

But i don't known how that works with api-platform.

File: config/services.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# This file is the entry point to configure your own services.
+# Files in the packages/ subdirectory configure your dependencies.
+
+# Put parameters here that don't need to change on each machine where the app is deployed
+# https://symfony.com/doc/current/best_practices/configuration.html#application-related-configuration
+parameters:
+    env(MONGODB_URL): ''
+    env(MONGODB_DB): ''
+    env(ELASTICSEARCH_URL): ''
```
Comment (cableman) [inline config/services.yaml:9]: How do this way relate to the way we do it in the other services: https://github.com/danskernesdigitalebibliotek/ddb-cover-service/blob/develop/config/services.yaml#L8

File: public/openapi/openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,166 @@
+openapi: 3.0.2
```
Comment (cableman) [inline public/openapi/openapi.yaml]: It this file required? I don't have the file as this changes can be set in https://github.com/danskernesdigitalebibliotek/ddb-cover-service-upload/blob/develop/config/packages/api_platform.yaml

File: public/openapi/style.css [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+/**  METHODS BLOCS **/
+
+#swagger-ui.api-platform .opblock-tag:hover {
+  background-color: inherit;
+  transform: scale(1.01);
+}
+
+#swagger-ui.api-platform .opblock.opblock-get .opblock-summary-method {
+  background-color:rgb(40, 143, 136);
+}
```
Comment (cableman) [inline public/openapi/style.css]: Missing newline :-)

File: src/Command/FakeCommand.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        $io = new SymfonyStyle($input, $output);
+
+        $date = $io->ask('Select a date (for example "7 december 2019" or "-2 days")?', 'today');
+
+        $date = new \DateTime($date);
+
+        $this->fakerService->createElasticsearchTestData($date);
+
+        $io->success('Adding fake data to elasticsearch successfully.');
```
Comment (cableman) [inline src/Command/FakeCommand.php:69]: Will it always be "successfully"

File: src/Document/Entry.php [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+ * @ApiResource(
+ *     collectionOperations={"get"},
+ *     itemOperations={"get"},
+ * )
+ * @ApiFilter(DateFilter::class, properties={"date"})
+ * @ApiFilter(BooleanFilter::class, properties={"extracted"})
+ *
+ * @ODM\Document
+ */
+class Entry implements \JsonSerializable
```
Comment (cableman) [inline src/Document/Entry.php]: How is this related to NormalizerInterface and https://symfony.com/blog/new-in-symfony-4-1-faster-serializer

File: src/Document/ExtractionResult.php [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace App\Document;
+
+use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;
+
+/**
+ * @ODM\Document(repositoryClass=App\Repository\ExtractionResultRepository::class)
+ */
+class ExtractionResult implements \JsonSerializable
+{
```
Comment (cableman) [inline src/Document/ExtractionResult.php:11]: Same comment about NormalizerInterface as the file above.

File: tests/Functional/FunctionalTest.php [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+        $output = new NullOutput();
+        $command->run($input, $output);
+        $result = $command->run($input, $output);
+        $this->assertEquals(0, $result, 'CleanupEntriesCommand should return 0');
+
+        // Test FaktCommand.
+        $command = new FakeCommand(new DataFakerService($httpClient, $documentManager, ''));
+        $command->getDescription();
+        $this->assertEquals('Add fake data to elasticsearch', $command->getDescription(), 'Description should have been set.');
+        // Will not test run, since it asks a question.
```
Comment (cableman) [inline tests/Functional/FunctionalTest.php]: ???

File: src/Service/StatisticsExtractionService.php [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+     * @throws \Doctrine\ODM\MongoDB\MongoDBException
+     * @throws \Throwable
+     */
+    public function extractStatistics()
+    {
+        // Get latest extraction entry. Default to first day of production.
+        /* @var ExtractionResult $lastExtraction */
+        $lastExtraction = $this->extractionResultRepository->getNewestEntry();
+
+        // @TODO: Set correct date of first stats entry
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php]: TODO?

File: src/Service/StatisticsExtractionService.php [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+        /* @var \DateTime $latestExtractionDate */
+        $latestExtractionDate = $lastExtraction ? $lastExtraction->getDate() : new \Datetime('1 december 2019');
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php:69]: The hardcoded date may need a comment.

File: src/Service/StatisticsExtractionService.php [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+        $today = new \DateTime();
+        $numberOfDaysToSearch = (int) $today->diff($latestExtractionDate)->format('%a');
+
+        $numberOfEntriesAdded = 0;
+
+        // Extract logs for all dates from latest extraction date to yesterday.
+        // Craete all as Entries in the database.
+        while ($numberOfDaysToSearch > 0) {
+            $dayToSearch = new \DateTime('-'.($numberOfDaysToSearch - 1).' days');
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php:78]: Should it be `$daysToSearch` as it looks into `$numberOfDaysToSearch`?

Comment (cableman) [reply src/Service/StatisticsExtractionService.php:78]: or is it -20 days, -19 days... so it's on day pr. loop?

File: src/Service/StatisticsExtractionService.php [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+                        $entry->setEvent('request_image');
+                        $entry->setResponse(json_encode($response));
+                        $entry->setExtracted(false);
+
+                        $this->documentManager->persist($entry);
+                        ++$numberOfEntriesAdded;
+                    }
+                } else {
+                    // Version 1 of statistics logging, where matches is not set.
+
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php]: Empty line after a line comment?

File: src/Service/StatisticsExtractionService.php [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+        }
+
+        // Save new extraction result.
+        $extractionResult = new ExtractionResult();
+        $extractionResult->setDate(new \DateTime());
+        $extractionResult->setNumberOfEntriesAdded($numberOfEntriesAdded);
+        $this->documentManager->persist($extractionResult);
+
+        // Flush to database.
+        $this->documentManager->flush();
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php]: Maybe flash a batch at the time to clear up memory... as the persist will keep stuff in memory until flush is called. Maybe event call `->clean()` as i remember.

File: src/Service/StatisticsExtractionService.php [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+            if (null !== $entry->getExtracted() && null !== $entry->getExtractionDate()) {
+                $diff = (int) $entry->getExtractionDate()->diff($compareDate)->format('%a');
+
+                if (0 < $diff && $entry->getExtractionDate() < $compareDate) {
+                    $this->documentManager->remove($entry);
+                }
+            }
+        }
+
+        $this->documentManager->flush();
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php:245]: Same here maybe flush for every 50-100 records to clear up memory.

File: src/Service/ElasticsearchServiceMock.php [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+ * @file
+ * Contains a mock of ElasticsearchService.
+ */
+
+namespace App\Service;
+
+/**
+ * Class ElasticsearchService.
+ */
+class ElasticsearchServiceMock extends ElasticsearchService
```
Comment (cableman) [inline src/Service/ElasticsearchServiceMock.php]: Is this mock service needed? If so I don't think it should be places with the normal services code or it should be placed inside test folder and namespaced under test.

Comment (cableman) [reply src/Service/ElasticsearchServiceMock.php]: Look at https://symfony.com/doc/4.4/create_framework/unit_testing.html and search for `createMock`

File: src/Service/DataFakerService.php [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace App\Service;
+
+use Doctrine\ODM\MongoDB\DocumentManager;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * Class ElasticsearchFaker.
+ */
+class DataFakerService
```
Comment (cableman) [inline src/Service/DataFakerService.php]: Should this service have been in a test namespace.

File: src/Repository/ExtractionResultRepository.php [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * ExtractionResultRepository constructor.
+     *
+     * @param \Doctrine\ODM\MongoDB\DocumentManager $documentManager
+     *   The doctrine document manager
+     */
+    public function __construct(DocumentManager $documentManager)
+    {
+        $uow = $documentManager->getUnitOfWork();
+        $classMetaData = $documentManager->getClassMetadata(ExtractionResult::class);
```
Comment (cableman) [inline src/Repository/ExtractionResultRepository.php:30]: This magic might require more comments.

File: src/EventSubscriber/ResponseSubscriber.php [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+
+        $extractionDatetime = new \DateTime();
+
+        // If requesting Event collection, mark each element as already retrieved.
+        if ('api_entries_get_collection' === $request->attributes->get('_route') &&
+            Entry::class === $request->attributes->get('_api_resource_class')) {
+            $results = json_decode($event->getResponse()->getContent());
+
+            // If it is an json:ld request.
+            if (isset($results->{'hydra:member'})) {
```
Comment (cableman) [inline src/EventSubscriber/ResponseSubscriber.php]: Think we should disable json:ld requests.

File: config/services.yaml [hunk 23]
```diff
@@ -0,0 +1,23 @@
+# This file is the entry point to configure your own services.
+# Files in the packages/ subdirectory configure your dependencies.
+
+# Put parameters here that don't need to change on each machine where the app is deployed
+# https://symfony.com/doc/current/best_practices/configuration.html#application-related-configuration
+parameters:
+    env(MONGODB_URL): ''
```
Comment (cableman) [inline config/services.yaml:7]: Why is it done this way vs. https://github.com/danskernesdigitalebibliotek/ddb-cover-service-importers/blob/develop/config/services.yaml#L8 ?

**Review by cableman — CHANGES_REQUESTED** [review]

File: public/openapi/style.css (hunk 8 above)
Comment (tuj) [reply public/openapi/style.css]: :D

File: src/Service/StatisticsExtractionService.php (hunk 15 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php:78]: It's one day at a time... will see if I can clarify

File: src/Service/StatisticsExtractionService.php (hunk 16 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php]: Removed

File: src/Document/Entry.php (hunk 10 above)
Comment (tuj) [reply src/Document/Entry.php]: This is just to expose private fields to json_encode.
ApiPlatform, Doctrine ODM handles normalization/serialization independent of json_encode/json_decode.
Since I have not written a Symfony normalizer, the above does not apply. The above optimization is applied automatically, unless you have written your own normalizer.

File: composer.json (hunk 1 above)
Comment (tuj) [reply composer.json]: Fixed

File: config/routes/api_platform.yaml (hunk 4 above)
Comment (tuj) [reply config/routes/api_platform.yaml]: Fixed

File: src/Command/FakeCommand.php (hunk 9 above)
Comment (tuj) [reply src/Command/FakeCommand.php:69]: If it reaches this point :)

File: src/Repository/ExtractionResultRepository.php (hunk 21 above)
Comment (tuj) [reply src/Repository/ExtractionResultRepository.php:30]: Fixed

File: config/packages/api_platform.yaml (hunk 2 above)
Comment (tuj) [reply config/packages/api_platform.yaml]: Fixed

File: tests/Functional/FunctionalTest.php (hunk 12 above)
Comment (tuj) [reply tests/Functional/FunctionalTest.php]: Fixed

File: src/Service/StatisticsExtractionService.php (hunk 13 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php]: Fixed

File: src/Service/StatisticsExtractionService.php (hunk 14 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php:69]: Added comment

File: src/Service/ElasticsearchServiceMock.php (hunk 19 above)
Comment (tuj) [reply src/Service/ElasticsearchServiceMock.php]: Fixed

File: src/Service/DataFakerService.php (hunk 20 above)
Comment (tuj) [reply src/Service/DataFakerService.php]: Fixed

File: src/EventSubscriber/ResponseSubscriber.php (hunk 22 above)
Comment (tuj) [reply src/EventSubscriber/ResponseSubscriber.php]: Fixed

File: config/packages/cache.yaml (hunk 3 above)
Comment (tuj) [reply config/packages/cache.yaml:2]: Fixed

File: src/Service/StatisticsExtractionService.php (hunk 17 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php]: Fixed

File: src/Service/StatisticsExtractionService.php (hunk 18 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php:245]: Fixed

File: config/services.yaml (hunk 23 above)
Comment (tuj) [reply config/services.yaml:7]: This is the recommended approach:
https://symfony.com/doc/current/configuration/env_var_processors.html#built-in-environment-variable-processors

File: src/Document/Entry.php (hunk 10 above)
Comment (tuj) [reply src/Document/Entry.php]: Removed since it was unused

File: src/Document/ExtractionResult.php (hunk 11 above)
Comment (tuj) [reply src/Document/ExtractionResult.php:11]: Removed since it was unused

File: config/services.yaml (hunk 6 above)
Comment (tuj) [reply config/services.yaml:9]: This is the recommended approach:
https://symfony.com/doc/current/configuration/env_var_processors.html#built-in-environment-variable-processors

File: config/packages/security.yaml (hunk 5 above)
Comment (tuj) [reply config/packages/security.yaml:12]: I don't this optimization will be necessary.

File: public/openapi/openapi.yaml (hunk 7 above)
Comment (tuj) [reply public/openapi/openapi.yaml]: Fixed

File: config/services.yaml (hunk 6 above)
Comment (cableman) [reply config/services.yaml:9]: As I see it we are using them for different things and both are metodes are listed in the link you link to.

File: .env [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+###> nelmio/cors-bundle ###
+CORS_ALLOW_ORIGIN='^https?://localhost(:[0-9]+)?$'
+###< nelmio/cors-bundle ###
+
+###> doctrine/doctrine-bundle ###
+# Format described at https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/configuration.html#connecting-using-a-url
+# For an SQLite database, use: "sqlite:///%kernel.project_dir%/var/data.db"
+# For a PostgreSQL database, use: "postgresql://db_user:db_password@127.0.0.1:5432/db_name?serverVersion=11&charset=utf8"
+# IMPORTANT: You MUST configure your server version, either here or in config/packages/doctrine.yaml
+DATABASE_URL=mysql://db_user:db_password@127.0.0.1:3306/db_name?serverVersion=5.7
```
Comment (cableman) [inline .env]: Normally "we/I" would set the defaults here that matches the docker-compose setup for database connection.

Comment (cableman) [reply .env]: If doctrine (normal database url) is not used by the site it should be possible to remove this completely from the codebase.

File: .github/workflows/pr.yaml [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+                with:
+                    php-version: 7.3
+                    extension-csv: ctype, iconv, json
+                    coverage: none
+            -   name: Validate composer files
+                run: |
+                    composer validate composer.json
+            -   name: Composer install with exported .env variables
+                run: |
+                    set -a && source .env && set +a
```
Comment (cableman) [inline .github/workflows/pr.yaml]: Why is this needed? Should composer/symfony not do this for you?

File: src/Document/ExtractionResult.php [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * @param mixed $date
+     */
+    public function setDate($date): void
+    {
+        $this->date = $date;
+    }
+
+    /**
+     * @return mixed
```
Comment (cableman) [inline src/Document/ExtractionResult.php]: Why is this `mixed` when the type is `int`?

File: src/Document/ExtractionResult.php [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public function getNumberOfEntriesAdded()
+    {
+        return $this->numberOfEntriesAdded;
+    }
+
+    /**
+     * @param mixed $numberOfEntriesAdded
```
Comment (cableman) [inline src/Document/ExtractionResult.php]: Why is this `mixed` when the type is `int`?

File: src/Document/Entry.php [hunk 28]
```diff
@@ -0,0 +1,253 @@
+<?php
```
Comment (cableman) [inline src/Document/Entry.php:1]: General comment: Why is all the parameter and return types `mixed` in these `src/Document/*` files?

File: src/Repository/ExtractionResultRepository.php (hunk 21 above)
Comment (cableman) [reply src/Repository/ExtractionResultRepository.php:30]: It's still black magic. But what do it do? UnitOfWork related to entity manager ?

File: src/Repository/EntryRepository.php [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+     *
+     * @param \Doctrine\ODM\MongoDB\DocumentManager $documentManager
+     *   The doctrine document manager
+     */
+    public function __construct(DocumentManager $documentManager)
+    {
+        // Because it cannot be auto-wired.
+        $uow = $documentManager->getUnitOfWork();
+        $classMetaData = $documentManager->getClassMetadata(Entry::class);
+        parent::__construct($documentManager, $uow, $classMetaData);
```
Comment (cableman) [inline src/Repository/EntryRepository.php:31]: It's still black magic. But what do it do? UnitOfWork related to entity manager ?

File: src/Service/ElasticsearchService.php [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * StatisticsExtractionService constructor.
+     *
+     * @param \Symfony\Contracts\HttpClient\HttpClientInterface $httpClient
+     *   The http client
+     * @param $boundElasticsearchURL
+     *   Url of elasticsearch instance
+     */
+    public function __construct(HttpClientInterface $httpClient, $boundElasticsearchURL)
```
Comment (cableman) [inline src/Service/ElasticsearchService.php]: Missing type for `$boundElasticsearchURL`.

File: src/Service/ElasticsearchService.php [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+     *
+     * @param \DateTime $date
+     *   Requested date
+     * @param string $message
+     *   The message to search for
+     *
+     * @return array
+     *   Array of logs for the given date
+     *
+     * @throws \Throwable
```
Comment (cableman) [inline src/Service/ElasticsearchService.php]: Would it not be an exception that it throws?

https://www.php.net/manual/en/class.throwable.php

File: src/Service/ElasticsearchService.php [hunk 32]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+/**
+ * @file
+ * Contains service to integrate with elasticsearch.
```
Comment (cableman) [inline src/Service/ElasticsearchService.php:5]: Found out the in Drupal 8 the file comment content is moved into the `@class` description above the class. Not sure what the standard is for symfony... and have not found a clear answer on google.

File: src/Service/StatisticsExtractionService.php [hunk 33]
```diff
@@ -0,0 +1,271 @@
...
+     */
+    public function extractStatistics()
+    {
+        $today = new \DateTime();
+
+        // Get latest extraction entry. Default to first day of production.
+        /* @var ExtractionResult $lastExtraction */
+        $lastExtraction = $this->extractionResultRepository->getNewestEntry();
+
+        // Default to 1. december 2019 to make sure we extract all statistics from the start of production.
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php]: We might need to add here that it's the production date for http://cover.dandigbib.org/ and not the faktor service.

File: src/Service/StatisticsExtractionService.php [hunk 34]
```diff
@@ -0,0 +1,271 @@
...
+
+            // Add all to mongodb.
+            foreach ($statistics as $statisticsEntry) {
+                // Flush when batch size is exceeded to avoid memory buildup.
+                if (0 < $numberOfEntriesAdded && $numberOfEntriesAdded > $nextBatchLimit) {
+                    $nextBatchLimit = $nextBatchLimit + self::BATCH_SIZE;
+                    $this->documentManager->flush();
+                    $this->documentManager->clear();
+
+                    gc_collect_cycles();
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php]: Why run the GC here?

File: src/Service/StatisticsExtractionService.php [hunk 35]
```diff
@@ -0,0 +1,271 @@
...
+
+            // Get entries from elasticsearch index.
+            $statistics = $this->elasticsearchService->getLogsFromElasticsearch($dayToSearch, 'Cover request/response');
+
+            $nextBatchLimit = self::BATCH_SIZE;
+
+            // Add all to mongodb.
+            foreach ($statistics as $statisticsEntry) {
+                // Flush when batch size is exceeded to avoid memory buildup.
+                if (0 < $numberOfEntriesAdded && $numberOfEntriesAdded > $nextBatchLimit) {
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php]: Could this be simplified if it was place at the end og the loop (then `$numberOfEntriesAdded` would properly never be `0`)

File: src/Service/StatisticsExtractionService.php (hunk 18 above)
Comment (cableman) [reply src/Service/StatisticsExtractionService.php:245]: This is still flushed outside the delete loop?

**Review by cableman — CHANGES_REQUESTED** [review]

File: .github/workflows/pr.yaml (hunk 25 above)
Comment (tuj) [reply .github/workflows/pr.yaml]: Not when running composer commands

Comment (cableman) [reply .github/workflows/pr.yaml]: Not sure I understand this... I don't do this when building docker images with Symfony applications. What is in `.env` thats needed during `composer install` here not normally required. It might be me that is missing something here...

Comment (tuj) [reply .github/workflows/pr.yaml]: Composer install runs cache:clear which invokes symfony. So if environment variables are not set, this will provoke an exception

Comment (tuj) [reply .github/workflows/pr.yaml]: I will try commenting out the line, to see if it is needed

File: src/Document/Entry.php (hunk 28 above)
Comment (tuj) [reply src/Document/Entry.php:1]: Because I just let phpstorm generate the getters and setter, since there is no equivalent of make:entity with Documents. So they are written by hand.
Will fix

File: src/Service/ElasticsearchService.php (hunk 32 above)
Comment (tuj) [reply src/Service/ElasticsearchService.php:5]: Will leave it as is atm.

File: .github/workflows/pr.yaml (hunk 25 above)
Comment (tuj) [reply .github/workflows/pr.yaml]: Fixed

File: src/Document/ExtractionResult.php (hunk 26 above)
Comment (tuj) [reply src/Document/ExtractionResult.php]: Fixed

File: src/Document/ExtractionResult.php (hunk 27 above)
Comment (tuj) [reply src/Document/ExtractionResult.php]: Fixed

File: src/Document/Entry.php (hunk 28 above)
Comment (tuj) [reply src/Document/Entry.php:1]: Fixed

File: src/Repository/EntryRepository.php (hunk 29 above)
Comment (tuj) [reply src/Repository/EntryRepository.php:31]: Have expanded comment

File: src/Service/ElasticsearchService.php (hunk 30 above)
Comment (tuj) [reply src/Service/ElasticsearchService.php]: Fixed

File: src/Service/ElasticsearchService.php (hunk 31 above)
Comment (tuj) [reply src/Service/ElasticsearchService.php]: Fixed

File: src/Service/StatisticsExtractionService.php (hunk 18 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php:245]: Fixed

File: src/Repository/ExtractionResultRepository.php (hunk 21 above)
Comment (tuj) [reply src/Repository/ExtractionResultRepository.php:30]: Have added bigger comment

File: src/Service/StatisticsExtractionService.php (hunk 34 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php]: Fixed

File: src/Service/StatisticsExtractionService.php (hunk 33 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php]: Fixed

File: src/Service/StatisticsExtractionService.php (hunk 35 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php]: Fixed.
I have added it to the start of the loop, because a variable number of entries are added for each iteration of the loop. This way the `if`-block is only added one place, instead of before each `continue`.

File: .env (hunk 24 above)
Comment (tuj) [reply .env]: Removed doctrine orm integration. But I still use Doctrine with regards to mongodb

File: src/Service/StatisticsExtractionService.php [hunk 36]
```diff
@@ -218,13 +222,22 @@ public function removeExtractedEntries(\DateTime $compareDate)
...
                 $diff = (int) $entry->getExtractionDate()->diff($compareDate)->format('%a');
 
                 if (0 < $diff && $entry->getExtractionDate() < $compareDate) {
                     $this->documentManager->remove($entry);
+
+                    $entriesRemoved++;
+
+                    // Flush when batch size is exceeded to avoid memory buildup.
+                    if ($entriesRemoved % self::BATCH_SIZE === 0) {
+                        $this->documentManager->flush();
```
Comment (cableman) [inline src/Service/StatisticsExtractionService.php:239]: What happens if the removed entries size is smaller than the batch size or the last batch is smaller?

**Review by cableman — CHANGES_REQUESTED** [review]

File: src/Service/StatisticsExtractionService.php (hunk 36 above)
Comment (tuj) [reply src/Service/StatisticsExtractionService.php:239]: Then the loop terminates and
$this->documentManager->flush();
is called at the end.

Comment (tuj) [reply src/Service/StatisticsExtractionService.php:239]: https://github.com/danskernesdigitalebibliotek/ddb-cover-service-faktor-export/pull/1/files/71aaceadc8606265f89225c4b92287472f9840c7..cc2e5532946b705f694ff4b76a1aa2ad1aacfcf1#diff-e8ce8b090fc1981cbba9e4709b0673f4R245

**Review by cableman — APPROVED** [review]
